### ISSUE TITLE
genpkey not working properly with (external) provider signature algorithm

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#14069]
This issue pertains to master and can be reproduced (using [oqsprovider](https://github.com/open-quantum-safe/oqs-provider)) running the following:

```
LD_LIBRARY_PATH=.local/lib .local/bin/openssl genpkey -provider-path _build/oqsprov  -provider oqsprovider -algorithm dilithium2
```

never returns but has the CPU spinning at 100%.

With parameters arranged differently, an equally unexpected error is reported:

```
>LD_LIBRARY_PATH=.local/lib .local/bin/openssl genpkey -algorithm dilithium2 -provider-path _build/oqsprov  -provider oqsprovider -algorithm dilithium2 
Error initializing dilithium2 context
C091DB79F37F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:331:Global default library context, Algorithm (dilithium2 : 0), Properties (<null>)
```

**Comment by t8m** [comment]
Could you please try to get some backtraces from the hung process?

The second one fails because putting -algorithm dilithium2 before the -provider does not work as the argument is processed immediately and the algorithm is unknown to the default provider. Algorithm order is in many cases significant in the openssl apps so I do not think this is an issue.

**Comment by levitte** [comment]
> The second one fails because putting -algorithm dilithium2 before the -provider d...

@baentsch, is there a reason you're having `-algorithm dilithium2` twice on that command line?

**Comment by levitte** [comment]
I just tried the exact same thing...  it doesn't spin, but I do get an error:

``` console
$ LD_LIBRARY_PATH=$HOME/.private-install/opt/openssl-3.0/lib/ $HOME/.private-install/opt/openssl-3.0/bin/openssl genpkey -provider-path _build/oqsprov  -provider oqsprovider -algorithm dilithium2
Algorithm dilithium2 not found
```

**Comment by mattcaswell** [comment]
I can recreate this. Investigating...

**Comment by mattcaswell** [comment]
Aha! There are two separate problems which are interacting badly to create this result.

First, you have an error on your command line. Because you are explicitly loading a provider we do not load the default provider automatically. Therefore if you *also* need default then you must explicitly load it. In other words add an additional "-provider default" to your command line.

The second problem is a bug in liboqs which I've just reported here:

https://github.com/open-quantum-safe/liboqs/issues/895

Basically liboqs calls OpenSSL's RAND_bytes() function in an infinite loop. If the call to RAND_bytes() fails it just keeps going on forever trying to get the random data. In this case the call failed because the default provider was not available, so there are no DRBGs to fetch.

**Comment by mattcaswell** [comment]
The second problem where you get the fetch failure is because genpkey seems to process the "-algorithm" option immediately as soon as its read the argument, i.e. before its processed the "-provider" options. We should probably defer processing "-algorithm" until later.

**Comment by levitte** [comment]
I rebuilt my OpenSSL 3 installation (it was a bit aged) and *now* I get the endless spin...  so this is a fairly recent problem

**Comment by mattcaswell** [comment]
> Algorithm order is in many cases significant in the openssl apps so I do not think this is an issue.

At the very least we should document that -algorithm needs to come after -provider. However, I think that is a fairly horrible design and will trip lots of people up. It's also new for this release since previous releases didn't have "-provider".

I would prefer that we should not require our users to do this in the right order. Instead we should remember the "-algorithm" selection and process it *after* the provider processing. However this argument ordering issue is bigger than just genpkey, so it may be too big a problem.

Either way it should be fixed in one way or another (just documentation or properly). But probably not in the beta1 timescale.

**Comment by richsalz** [comment]
I would be willing to do the "hold algorithm until providers are loaded" work if you think it's necessary for 3.0

**Comment by mattcaswell** [comment]
> I would be willing to do the "hold algorithm until providers are loaded" work if you think it's necessary for 3.0

That would be great. I think it would be a significant usability benefit for it to be fixed for 3.0...although perhaps not *absolutely* necessary (i.e. we shouldn't necessarily hold 3.0 if we didn't fix it).

As mentioned above there is probably a much bigger problem across all the apps....

**Comment by baentsch** [comment]
>  However, I think that is a fairly horrible design and will trip lots of people up.

From the perspective of a "mere user" I whole-heartedly agree. It would be good to fix before a beta release signals usability readiness to many more people (and _most_ of them will not read a doc as I guess we all know).

**Comment by richsalz** [comment]
> As mentioned above there is probably a much bigger problem across all the apps....

I'll fix the algorithm in speed and genpkey.  Aliases for "enc" name will be too hard, but I could do the `-md` and `-cipher` flags which is about 15 files.  Any other specific cases?

**Comment by baentsch** [comment]
> Any other specific cases?

This all got triggered by the goal to do [PQC CMS](https://github.com/open-quantum-safe/openssl#cms-demo) now also via a provider, so the full command set will probably consist of `genpkey`, `[x509] -req` and `cms`. Overall, we also regularly use `speed` and `s_time` (beyond `s_client` and `s_server` that already seem to function properly -- without having twiddled parameters there yet...).

**Comment by mattcaswell** [comment]
> This all got triggered by the goal to do PQC CMS now also via a provider

My guess is you will run into some problems due to this code:

https://github.com/openssl/openssl/blob/88444854affe31ce08a5daaf4b6afc86e6972c63/crypto/cms/cms_env.c#L118-L129

i.e. some hard coded assumptions about the types of CMS key that we support.

Similarly here:
https://github.com/openssl/openssl/blob/88444854affe31ce08a5daaf4b6afc86e6972c63/crypto/cms/cms_sd.c#L235-L244

**Comment by mattcaswell** [comment]
> Any other specific cases?

My guess is the "-paramfile" option to genpkey is similarly affected. 

Anything with a "-rand" option could end up fetching the DRBG from available providers.

There are some files that process flags using "opt_md" but the option isn't called "-md" - so I'd search for use of "opt_md", "opt_cipher", etc.

Beyond that the "bigger problem" I mentioned was referring to a larger issue of interdependencies between different flags generally (not just in the case of the "-provider" flags)...but I just had a quick scan through the apps and failed to find much. So perhaps not such a big issue as I thought - with perhaps "genpkey" being one where this does occur.

**Comment by richsalz** [comment]
@paulidale does `RAND_load_file()` depend on the provider?

**Comment by paulidale** [comment]
> @paulidale does `RAND_load_file()` depend on the provider?

Yes it will.  The `RAND_status()` call at the very end will cause the primary DRBG to be created.

This could be changed, `RAND_load_file()` is a otherwise NOP if a RAND_METHOD has not been installed.

**Comment by richsalz** [comment]
> This could be changed, RAND_load_file() is a otherwise NOP if a RAND_METHOD has not been installed.

I think it will be easier, but admittedly not cleaner, if I just handle this is in the apps for now.  Thanks!

**Comment by richsalz** [comment]
@baentsch I think I've fixed the issues in #14135, please take a look if you can.

**Comment by romen** [comment]
Argument order trips users in a similar way in 1.1.1 for engines provided on the command line: should that change in master also for engines?
While discussing the order of arguments, what about `-provider-path`? Does it support being issued more than once or a separator to signal an array of paths? Should the relative order of `-provider` and `-provider-path` options influence the outcome of loading providers?

Similar questions arise for the `-propquery` argument on the command line: if we are changing things so that the argument order is not significant anymore, we should probably delay fetching parsing any argument until all the provider and engine options, and we should probably apply this in all the apps in a systematic way rather than piecemeal for a subset of the apps.

Bonis questions: what about `-config`? Should we do several argument parsing passes and prioritize `-config` even before provider and engine options? What if `-config` is repeated as a cli argument?

**Comment by paulidale** [comment]
`-provider-path` does not (currently) support multiple directories and the relative of this and `-provider` is important.

**Comment by baentsch** [comment]
> @baentsch I think I've fixed the issues in #14135, please take a look if you can.

Done & confirmed: The "argument ordering issue" reported above is gone with #14135 applied and reproducible if not. Thanks very much!

**Comment by richsalz** [comment]
@romen your comment is important; can you open it as a separate issue for the project to look at?

**Comment by baentsch** [comment]
> should probably apply this in all the apps in a systematic way rather than piecemeal for a subset of the apps.

When glancing over #14135, the sheer number of files touched made me wonder something similar. But then again, there's the 3.0 beta goal... But having an issue for this as suggested by @richsalz makes a lot of sense.

**Comment by richsalz** [comment]
IMO, the new 3.0 provider options make the order-dependant a much bigger deal than it is in 1.1

**Comment by baentsch** [comment]
> The second problem is a bug in liboqs which I've just reported here:
> open-quantum-safe/liboqs#895

For whatever it's worth, that one has been closed by https://github.com/open-quantum-safe/liboqs/pull/897. So, as far as this issue is concerned, I'd think it could be closed with #14135 landing (plus a longer-term issue as per the above). 
Thanks again @mattcaswell for the analysis and @richsalz for #14135.

**Comment by baentsch** [comment]
Closed as per the fixes above.
